File: C Code/mcts_iot.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>

#define K 24
#define B_START 20
#define B_MAX 30
#define B_MIN 10
#define C sqrt(2)




static const int E [] = {3, 2, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 2, 3, 4, 5, 5, 6, 6, 6, 6, 5, 5, 4};

typedef struct task {
    int id;
    int cost;
    int quality;
} task_t;

static const task_t tasks[] = {
    {.id = 1, .cost = 3, .quality = 5},
    {.id = 2, .cost = 2, .quality = 3},
    {.id = 3, .cost = 4, .quality = 6},
    {.id = 4, .cost = 8, .quality = 10},
    {.id = 5, .cost = 1, .quality = 1}
};

#define NUM_TASKS sizeof(tasks) / sizeof(task_t)

typedef struct edge {
    int task_idx;
    struct node *parent;
    struct node *child;
} edge_t;

typedef struct node {
    int battery_level;
    int timeslot;
    struct edge_list *parents;
    struct edge_list *children;
    int visits;
    int win_quality;
    bool possible_tasks[NUM_TASKS];
} node_t;

typedef struct edge_list {
    edge_t **list;
    int size;
} edge_list_t;

static node_t *nodes[K][B_MAX - B_MIN + 1] = { NULL };

static void die(char *msg) {
    perror(msg);
    exit(EXIT_FAILURE);
}

static void append_edge(edge_list_t **list, edge_t *element) {
    if(*list == NULL) {
        *list = (edge_list_t *) malloc(sizeof(edge_list_t));
        if(NULL == *list) {
            die("malloc");
        }
        (*list)->list = (edge_t **) malloc(sizeof(edge_t *));
        if(NULL == (*list)->list) {
            die("malloc");
        }
        (*list)->list[0] = element;
        (*list)->size = 1;
    } else {
        edge_t **new_list = (edge_t **) realloc((*list)->list, ((*list)->size + 1) * sizeof(edge_t *));
        if(NULL == new_list) {
            die("realloc");
        }
        (*list)->list = new_list;
        (*list)->list[(*list)->size] = element;
        (*list)->size++;
    }
}

static void get_possible_tasks(node_t *node, bool possible_tasks[NUM_TASKS]);

static node_t *create_node(int battery_level, int timeslot, edge_t *parent) {
    node_t *node = (node_t *) malloc(sizeof(node_t));
    if(NULL == node) {
        die("malloc");
    }
    node->battery_level = battery_level;
    node->timeslot = timeslot;
    node->visits = 0;
    node->win_quality = 0;
    node->children = NULL;
    node->parents = NULL;
    bool pos_tasks[NUM_TASKS];
    memset(pos_tasks, false, sizeof(pos_tasks)); 
    get_possible_tasks(node, pos_tasks);
    memcpy(node->possible_tasks, pos_tasks, sizeof(pos_tasks));
    if(NULL != parent) {
        append_edge(&node->parents, parent);
    }
    return node;
}

static edge_t *create_edge(int task, node_t *parent, node_t *child) {
    edge_t *edge = (edge_t *) malloc(sizeof(edge_t));
    if(NULL == edge) {
        die("malloc");
    }
    edge->task_idx = task;
    edge->parent = parent;
    edge->child = child;
    return edge;
}

// Node functions

static void get_possible_tasks(node_t *node, bool possible_tasks[NUM_TASKS]) {
    if(node->timeslot == K) {
        return;
    }
    for(int i = 0; i < NUM_TASKS; i++) {
        if(node->battery_level + E[node->timeslot] - tasks[i].cost >= B_MIN) {
            possible_tasks[i] = true;
        }
    }
}

static bool is_fully_expanded(node_t *node) {
    for(int i = 0; i < NUM_TASKS; i++) {
        if(node->possible_tasks[i]) {
            return false;
        }
    }
    return true;
}

static bool is_terminal(node_t *node) {
    if(node->timeslot == K) {
        return true;
    }
    bool pos_tasks[NUM_TASKS];
    memset(pos_tasks, false, sizeof(pos_tasks)); 
    get_possible_tasks(node, pos_tasks);
    for(int i = 0; i < NUM_TASKS; i++) {
        if(pos_tasks[i]) {
            return false;
        }
    }
    return true;
}

static edge_t *expand(node_t *node) {
    int next_task_idx = -1;
    int new_battery = 0;
    bool new_child = false;

    // look for unexplored children
    while (!is_fully_expanded(node)) {
        for(int i = 0; i < NUM_TASKS; i++) {
            if(node->possible_tasks[i]) {
                node->possible_tasks[i] = false;
                next_task_idx = i;
                break;
            }
        }
        new_battery = node->battery_level + E[node->timeslot] - tasks[next_task_idx].cost;
        if(new_battery > B_MAX) {
            new_battery = B_MAX;
        }
        if(NULL != nodes[node->timeslot][new_battery - B_MIN]) {
            edge_t *edge = create_edge(next_task_idx, node, nodes[node->timeslot][new_battery - B_MIN]);
            append_edge(&(node->children), edge);
            append_edge(&(nodes[node->timeslot][new_battery - B_MIN]->parents), edge);
        } else {
            new_child = true;
            break;
        }
    }

    if(is_fully_expanded(node) && !new_child) {
        return NULL;
    }
    
    edge_t *edge = create_edge(next_task_idx, node, NULL);
    edge->child = create_node(new_battery, node->timeslot + 1, edge);
    nodes[node->timeslot][new_battery - B_MIN] = edge->child;
    append_edge(&(node->children), edge);
    return edge;
}

static edge_t *get_best_move(node_t *node) {
    float max_val = -1;
    edge_t *max_edge = NULL;
    for(int i = 0; i < node->children->size; i++) {
        node_t *child = node->children->list[i]->child;
        float val = ((float) child->win_quality / child->visits) + C * sqrt(log(node->visits) / child->visits);
        if(val > max_val) {
            max_val = val;
            max_edge = node->children->list[i];
        }
    }
    return max_edge;
}

static int simulate(node_t *node, int qual, int res[K]) {
    int sim_timeslot = node->timeslot;
    int sim_quality = qual;
    int sim_state_battery = node->battery_level;
    
    while(true) {
        if(K == sim_timeslot) {
            if(sim_state_battery < B_MIN) {
                sim_quality = 0;
            }
            return sim_quality;
        }

        // get random task
        int available_tasks[NUM_TASKS] = { 0 };
        int count = 0;
        for(int i = 0; i < NUM_TASKS; i++) {
            if(sim_state_battery + E[sim_timeslot] - tasks[i].cost >= B_MIN) {
                available_tasks[count] = i;
                count++;
            }
        }
        if(0 == count) {
            memset(res, 0, K * sizeof(int));
            return 0;
        }
        int random_idx = rand() % count;
        int chosen_task_idx = available_tasks[random_idx];

        // update simulation
        sim_quality += tasks[chosen_task_idx].quality;
        sim_state_battery = sim_state_battery + E[sim_timeslot] - tasks[chosen_task_idx].cost;
        if(sim_state_battery > B_MAX) {
            sim_state_battery = B_MAX;
        }
        res[sim_timeslot] = tasks[chosen_task_idx].id;
        sim_timeslot += 1;
    }
}

static void backpropagate(int result, node_t **path, int path_length) {
    for(int i = 0; i < path_length; i++) {
        path[i]->visits += 1;
        path[i]->win_quality += result;
    }
}

static int mcts(node_t *root, int iterations, int best_path[K]) {
    int best_quality = 0;

    for(int i = 0; i < iterations; i++) {
        node_t *node = root;
        edge_t *result = NULL;

        node_t *path[K + 1] = { NULL };
        path[0] = root;
        int path_length = 1;

        int task_path[K] = { 0 };
        int path_quality = 0;
        
        while (NULL == result) {
            // select
            while(!is_terminal(node) && is_fully_expanded(node)) {
                edge_t *best_move = get_best_move(node);
                node = best_move->child;
                path[path_length] = node;
                task_path[path_length - 1] = tasks[best_move->task_idx].id;
                path_length++;
                path_quality += tasks[best_move->task_idx].quality;
            }

            // expand
            if(!is_terminal(node)) {
                result = expand(node);
            } else {
                break;
            }
        }
        if(NULL != result) {
            node = result->child;
            path[path_length] = node;
            task_path[path_length - 1] = tasks[result->task_idx].id;
            
            path_length++;
            path_quality += tasks[result->task_idx].quality;
        }
        // simulate
        int res[K] = { 0 };
        int sim_quality = simulate(node, path_quality, res);
        if(sim_quality > best_quality) {
            best_quality = sim_quality;
            // concatenate paths to get full path
            for(int j = 0; j < K; j++) {
                if(task_path[j] != 0) {
                    best_path[j] = task_path[j];
                } else {
                    best_path[j] = res[j];
                }
            }
        }
        backpropagate(sim_quality, path, path_length);
    }
    return best_quality;
}

int main(int argc, char **argv) {
    srand(time(NULL));
    int best_path[K] = { 0 };
    node_t *root = create_node(B_START, 0, NULL);
    
    
    int ret = mcts(root, 500, best_path);
    int q = 0;
    printf("%d\n", ret);
    for(int i = 0; i < K; i++) {
        printf("%d\n", best_path[i]);
        q += tasks[best_path[i] - 1].quality;
    }
    return 0;
}

File: mcts_iot.py
import math
import random
import statistics
from collections import deque
import plotly.express as px
import numpy as np
import pandas as pd
from pyvis.network import Network


class Edge:
    def __init__(self, parent, child, task):
        self.parent = parent
        self.child = child
        self.task = task
        self.visits = 0
        self.win_quality = 0


class Node:
    def __init__(self, battery, timeslot, parent=None):
        self.battery = battery
        self.timeslot = timeslot  # timeslot 1 meaning from 0 to 1
        self.parents = [parent]
        self.children = []
        self.possible_tasks = self.get_possible_tasks()

    def get_possible_tasks(self):
        return [] if self.timeslot == K \
            else list(filter(lambda t: self.battery + E[self.timeslot] - t["cost"] >= B_min, Tasks))

    def is_fully_expanded(self):
        return len(self.possible_tasks) == 0

    def is_terminal(self):
        return len(self.get_possible_tasks()) == 0 or self.timeslot == K

    # returns: created edge and information if a new node was created => selection can be stopped
    def expand(self):
        # look for unexplored children
        next_task = self.possible_tasks.pop()
        new_battery = min(B_max, self.battery + E[self.timeslot] - next_task["cost"])
        if nodes[self.timeslot, new_battery - B_min] is not None:
            # child already exists, add edge
            edge = Edge(self, nodes[self.timeslot, new_battery - B_min], next_task)
            self.children.append(edge)
            nodes[self.timeslot, new_battery - B_min].parents.append(edge)
            return edge, False
        else:
            # child doesn't yet exist, add it
            edge = Edge(self, None, next_task)
            edge.child = Node(new_battery, self.timeslot + 1, parent=edge)
            nodes[self.timeslot, new_battery - B_min] = edge.child
            self.children.append(edge)
            return edge, True

    def get_best_move(self, parent_edge_visits, c=math.sqrt(2)):
        try:
            return max(self.children, key=lambda edge: (edge.win_quality / edge.visits) + c *
                                                   math.sqrt(math.log(parent_edge_visits) / edge.visits))
        except ZeroDivisionError:
            print(len(list(filter(lambda e: e.visits == 0, self.children))))
            print(self.timeslot)
            exit(1)


    def simulate(self, qual):
        sim_timeslot = self.timeslot
        sim_state_battery = self.battery
        sim_state_quality = qual
        sim_path = []
        while True:
            if sim_timeslot == K:
                return (sim_state_quality, sim_path, sim_state_quality, sim_state_battery) if sim_state_battery >= B_start else \
                    (penalized_quality(sim_state_quality, sim_state_battery), [], 0, 0)

            available_tasks = list(filter(lambda t: sim_state_battery + E[sim_timeslot] -
                                                    t["cost"] >= B_min, Tasks))
            if not available_tasks:
                return 0, [], 0, 0
            chosen_task = random.choice(available_tasks)
            sim_timeslot += 1
            sim_state_battery = min(B_max, sim_state_battery - chosen_task["cost"] + E[sim_timeslot - 1])
            sim_state_quality += chosen_task["quality"]
            sim_path.append(chosen_task)


def backpropagate(result, path):
    for edge in path:
        edge.visits += 1
        scaled_result = result
        edge.win_quality += scaled_result


def mcts(start_node, iterations=500):
    root = start_node

    # save the best path explored by selection, expansion and simulation
    best_path = []
    best_quality = 0
    best_path_remaining_battery = 0
    for j in range(iterations):
        node = root
        result = None
        # select until expand creates a new node
        # also memorize chosen path for backpropagation
        path = []
        task_path = []
        path_quality = 0
        new_node = False
        parent_edge_visits = j
        while not new_node:
            # select, create path
            while not node.is_terminal() and node.is_fully_expanded():
                best_move = node.get_best_move(parent_edge_visits)
                node = best_move.child
                parent_edge_visits = best_move.visits
                path.append(best_move)
                task_path.append(best_move.task)
                path_quality += best_move.task["quality"]
            # expand
            if not node.is_terminal():
                result, new_node = node.expand()
                node = result.child
                path.append(result)
                task_path.append(result.task)
                path_quality += result.task["quality"]
            else:
                break



        # simulate
        res, sim_path, sim_quality, sim_battery = node.simulate(path_quality)
        if sim_quality > best_quality:
            best_path = task_path + sim_path
            best_quality = sim_quality
            best_path_remaining_battery = sim_battery
        # backpropagate
        backpropagate(res, path)

    return root, best_path, best_quality, best_path_remaining_battery


def penalized_quality(quality, B_lvl):
    k = 1.5
    diff = max(0, B_start - B_lvl)  # only penalize if below start
    scale = diff / B_min
    penalty = quality * (1 - math.exp(-k * scale))
    return quality - penalty


Tasks = [{'id': 1, 'cost': 3, 'quality': 5},
         {'id': 2, 'cost': 2, 'quality': 3},
         {'id': 3, 'cost': 4, 'quality': 6},
         {'id': 4, 'cost': 8, 'quality': 10},
         {'id': 5, 'cost': 1, 'quality': 1}]
B_start = 20
B_max = 30
B_min = 10

E = [3, 2, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 2, 3, 4, 5, 5, 6, 6, 6, 6, 5, 5, 4]
K = 24


nodes = np.full((K, B_max + 1 - B_min), None, dtype=object)

'''

def evaluate(iterations=100):
    battery_values = []
    quality_values = []
    battery_underflow = 0
    errors = 0
    no_full_path = 0
    for i in range(iterations):
        global nodes
        nodes = np.full((K, B_max + 1 - B_min), None, dtype=object)
        curr_node = Node(B_start, 0)
        error = False
        solution = None
        try:
            solution = mcts(curr_node, 200)
        except ValueError:
            errors += 1
            error = True
        curr_node = solution[0]
        if not error:
            quality = 0
            while len(curr_node.children) != 0:
                move = curr_node.get_best_move()
                quality += move.task["quality"]
                curr_node = move.child
            battery_values.append(solution[3])
            quality_values.append(solution[2])
            if battery_values[-1] < B_start:
                battery_underflow += 1
            if curr_node.timeslot < 24:
                no_full_path += 1

    print(statistics.fmean(battery_values), statistics.fmean(quality_values), statistics.stdev(quality_values))


evaluate(1000)
'''

def eval_iterations():
    x_axis = []
    y_axis = []
    for i in range(10, 501, 10):
        x_axis.append(i)
        battery_values = []
        quality_values = []
        battery_underflow = 0
        errors = 0
        no_full_path = 0
        for j in range(100):
            global nodes
            nodes = np.full((K, B_max + 1 - B_min), None, dtype=object)
            curr_node = Node(B_start, 0)
            solution = mcts(curr_node, iterations=i)
            quality_values.append(solution[2])
            battery_values.append(solution[3])
            if battery_values[-1] < B_start:
                battery_underflow += 1
        y_axis.append(statistics.fmean(quality_values))
    fig = px.scatter(x=x_axis, y=y_axis)
    fig.show()



def visualize_tree(root, max_nodes=1000):
    net = Network(height='800px', width='100%', directed=True)
    queue = deque([root])
    visited = set()
    count = 0

    while queue and count < max_nodes:
        current = queue.popleft()
        node_id = str(id(current))
        if node_id in visited:
            continue
        visited.add(node_id)
        count += 1

        # node label
        label = f"T{current.timeslot}\nB{current.battery}"
        net.add_node(node_id, label=label, title=label, color="#a6cee3")

        for edge in current.children:
            child = edge.child
            if not child:
                continue
            child_id = str(id(child))
            label_text = f"Task {edge.task['id']}\nQ={edge.task['quality']}"
            color = "#1f78b4"

            # add child node if not seen yet
            if child_id not in visited:
                net.add_node(
                    child_id,
                    label=f"T{child.timeslot}\nB{child.battery}",
                    title=f"Battery={child.battery}, Timeslot={child.timeslot}",
                    color="#b2df8a"
                )
                queue.append(child)

            # Add edge with quality label
            net.add_edge(node_id, child_id, label=label_text, color=color, title=f"Quality={edge.task['quality']}")
    net.write_html("tree.html")


curr_node = Node(B_start, 0)
resu, bp, bq, bpb = mcts(curr_node, 150)
quality = 0
par_edge_visits = 150
while len(curr_node.children) != 0:
    move = curr_node.get_best_move(par_edge_visits)
    print(curr_node.timeslot, curr_node.battery, quality, move.task)
    par_edge_visits = move.visits
    quality += move.task["quality"]
    curr_node = move.child


print(curr_node.timeslot, curr_node.battery, quality)
print(bp, bq, bpb, len(bp))
'''
eval_iterations()
'''

